Application fails to start in Kubernetes when using target-class based (CGLIB) proxies (e.g., with EnableAsync)

**Describe the bug**
Hi there,

I'm currently using Spring Cloud Kubernetes, but run into an application startup issue after attempting to use an asynchronous method.
A sample Maven project to help reproduce the issue can be found in this repo : [rm3l/sck-issue-with-target-class-proxies](https://github.com/rm3l/sck-issue-with-target-class-proxies)

The error message (stack trace below) explicitly tells me to `Either pull the async method up to an interface or switch to CGLIB proxies by enforcing proxy-target-class mode`.
I don't understand why, since the use of CGLIB proxies is already being enforced application-wide (via the .`spring.aop.proxy-target-class` [Spring Boot property](https://docs.spring.io/spring-boot/docs/current/reference/html/appendix-application-properties.html#spring.aop.proxy-target-class) and the `proxyTargetClass=true` attribute of the `EnableAsync` annotation) 

I am not sure though if this is an issue related to Spring Cloud Kubernetes itself, or if it relates to Spring Boot instead.
Maybe I missed something, but the application starts and works like a charm when started outside of Kubernetes or when I explicitly set `spring.cloud.kubernetes.enabled=false`, which makes me suspect an issue with Spring Cloud Kubernetes.

Is there anything in Spring Cloud Kubernetes forcing the use of standard Java interface-based proxies, irrespective of how the application is configured ?

Below is an excerpt from the error log:
```
Error starting ApplicationContext. To display the conditions report re-run your application with 'debug' enabled.
2021-05-02 20:38:53.179 ERROR 1 --- [           main] o.s.boot.SpringApplication               : Application run failed

org.springframework.beans.factory.BeanInitializationException: Failed to process @EventListener annotation on bean with name 'testServiceImpl'; nested exception is java.lang.IllegalStateException: Need to invoke method 'handleEvent' declared on target class 'TestServiceImpl', but not found in any interface(s) of the exposed proxy type. Either pull the method up to an interface or switch to CGLIB proxies by enforcing proxy-target-class mode in your configuration.
	at org.springframework.context.event.EventListenerMethodProcessor.afterSingletonsInstantiated(EventListenerMethodProcessor.java:157) ~[spring-context-5.3.6.jar:5.3.6]
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:963) ~[spring-beans-5.3.6.jar:5.3.6]
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:918) ~[spring-context-5.3.6.jar:5.3.6]
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:583) ~[spring-context-5.3.6.jar:5.3.6]
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:144) ~[spring-boot-2.4.5.jar:2.4.5]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:782) ~[spring-boot-2.4.5.jar:2.4.5]
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:774) ~[spring-boot-2.4.5.jar:2.4.5]
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:439) ~[spring-boot-2.4.5.jar:2.4.5]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:339) ~[spring-boot-2.4.5.jar:2.4.5]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1340) ~[spring-boot-2.4.5.jar:2.4.5]
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1329) ~[spring-boot-2.4.5.jar:2.4.5]
	at org.rm3l.springcloudkubernetesissuewithtargetclassproxies.SpringCloudKubernetesIssueWithTargetClassProxiesApplication.main(SpringCloudKubernetesIssueWithTargetClassProxiesApplication.java:12) ~[classes/:na]
Caused by: java.lang.IllegalStateException: Need to invoke method 'handleEvent' declared on target class 'TestServiceImpl', but not found in any interface(s) of the exposed proxy type. Either pull the method up to an interface or switch to CGLIB proxies by enforcing proxy-target-class mode in your configuration.
	at org.springframework.core.MethodIntrospector.selectInvocableMethod(MethodIntrospector.java:132) ~[spring-core-5.3.6.jar:5.3.6]
	at org.springframework.aop.support.AopUtils.selectInvocableMethod(AopUtils.java:135) ~[spring-aop-5.3.6.jar:5.3.6]
	at org.springframework.context.event.EventListenerMethodProcessor.processBean(EventListenerMethodProcessor.java:198) ~[spring-context-5.3.6.jar:5.3.6]
	at org.springframework.context.event.EventListenerMethodProcessor.afterSingletonsInstantiated(EventListenerMethodProcessor.java:154) ~[spring-context-5.3.6.jar:5.3.6]
	... 11 common frames omitted
```

For reference, I am using the following versions:
- Spring Boot : 2.4.5
- Spring Cloud Kubernetes : 2.0.2

**Sample**
I've created this sample Maven project to help reproduce the issue: https://github.com/rm3l/sck-issue-with-target-class-proxies

This repo contains a simple Helm Chart for deploying to Kubernetes. It also contains a GitHub Workflow that tests this Chart. This job currently does not pass due to this issue.
Please do let me know if there is anything missing or if I can help further investigate.
Thanks.

--- a/spring-cloud-kubernetes-config/pom.xml
+++ b/spring-cloud-kubernetes-config/pom.xml
@@ -51,12 +51,14 @@
 			<groupId>org.springframework.boot</groupId>
 			<artifactId>spring-boot-autoconfigure</artifactId>
 		</dependency>
-
 		<dependency>
 			<groupId>org.springframework.cloud</groupId>
 			<artifactId>spring-cloud-context</artifactId>
 		</dependency>
-
+		<dependency>
+			<groupId>org.springframework.cloud</groupId>
+			<artifactId>spring-cloud-commons</artifactId>
+		</dependency>
 		<dependency>
 			<groupId>org.springframework.security</groupId>
 			<artifactId>spring-security-rsa</artifactId>

--- a/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java
+++ b/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java
@@ -30,6 +30,7 @@ import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
+import org.springframework.boot.task.TaskSchedulerBuilder;
 import org.springframework.cloud.autoconfigure.RefreshAutoConfiguration;
 import org.springframework.cloud.autoconfigure.RefreshEndpointAutoConfiguration;
 import org.springframework.cloud.context.refresh.ContextRefresher;
@@ -41,8 +42,7 @@ import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Conditional;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.core.env.AbstractEnvironment;
-import org.springframework.scheduling.annotation.EnableAsync;
-import org.springframework.scheduling.annotation.EnableScheduling;
+import org.springframework.scheduling.TaskScheduler;
 import org.springframework.util.Assert;
 
 /**
@@ -65,8 +65,6 @@ public class ConfigReloadAutoConfiguration {
 	 */
 	@ConditionalOnProperty("spring.cloud.kubernetes.reload.enabled")
 	@ConditionalOnClass({ RestartEndpoint.class, ContextRefresher.class })
-	@EnableScheduling
-	@EnableAsync
 	protected static class ConfigReloadAutoConfigurationBeans {
 
 		@Autowired
@@ -87,12 +85,14 @@ public class ConfigReloadAutoConfiguration {
 				@Autowired(
 						required = false) ConfigMapPropertySourceLocator configMapPropertySourceLocator,
 				@Autowired(
-						required = false) SecretsPropertySourceLocator secretsPropertySourceLocator) {
+						required = false) SecretsPropertySourceLocator secretsPropertySourceLocator,
+				TaskScheduler taskScheduler) {
 			switch (properties.getMode()) {
 			case POLLING:
 				return new PollingConfigurationChangeDetector(this.environment,
 						properties, this.kubernetesClient, strategy,
-						configMapPropertySourceLocator, secretsPropertySourceLocator);
+						configMapPropertySourceLocator, secretsPropertySourceLocator,
+						taskScheduler, properties);
 			case EVENT:
 				return new EventBasedConfigurationChangeDetector(this.environment,
 						properties, this.kubernetesClient, strategy,
@@ -100,6 +100,12 @@ public class ConfigReloadAutoConfiguration {
 			}
 			throw new IllegalStateException(
 					"Unsupported configuration reload mode: " + properties.getMode());
+		}
+
+		@Bean
+		@ConditionalOnMissingBean
+		public TaskScheduler taskScheduler() {
+			return new TaskSchedulerBuilder().build();
 		}
 
 		/**

--- a/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java
+++ b/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/ConfigReloadAutoConfiguration.java
@@ -30,9 +30,9 @@ import org.springframework.boot.autoconfigure.condition.ConditionalOnClass;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnMissingBean;
 import org.springframework.boot.autoconfigure.condition.ConditionalOnProperty;
 import org.springframework.boot.context.properties.EnableConfigurationProperties;
-import org.springframework.boot.task.TaskSchedulerBuilder;
 import org.springframework.cloud.autoconfigure.RefreshAutoConfiguration;
 import org.springframework.cloud.autoconfigure.RefreshEndpointAutoConfiguration;
+import org.springframework.cloud.commons.util.TaskSchedulerWrapper;
 import org.springframework.cloud.context.refresh.ContextRefresher;
 import org.springframework.cloud.context.restart.RestartEndpoint;
 import org.springframework.cloud.kubernetes.config.ConfigMapPropertySourceLocator;
@@ -42,7 +42,7 @@ import org.springframework.context.annotation.Bean;
 import org.springframework.context.annotation.Conditional;
 import org.springframework.context.annotation.Configuration;
 import org.springframework.core.env.AbstractEnvironment;
-import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
 import org.springframework.util.Assert;
 
 /**
@@ -86,13 +86,13 @@ public class ConfigReloadAutoConfiguration {
 						required = false) ConfigMapPropertySourceLocator configMapPropertySourceLocator,
 				@Autowired(
 						required = false) SecretsPropertySourceLocator secretsPropertySourceLocator,
-				TaskScheduler taskScheduler) {
+				TaskSchedulerWrapper taskScheduler) {
 			switch (properties.getMode()) {
 			case POLLING:
 				return new PollingConfigurationChangeDetector(this.environment,
 						properties, this.kubernetesClient, strategy,
 						configMapPropertySourceLocator, secretsPropertySourceLocator,
-						taskScheduler, properties);
+						taskScheduler.getTaskScheduler(), properties);
 			case EVENT:
 				return new EventBasedConfigurationChangeDetector(this.environment,
 						properties, this.kubernetesClient, strategy,
@@ -102,10 +102,16 @@ public class ConfigReloadAutoConfiguration {
 					"Unsupported configuration reload mode: " + properties.getMode());
 		}
 
-		@Bean
+		@Bean("springCloudKubernetesTaskScheduler")
 		@ConditionalOnMissingBean
-		public TaskScheduler taskScheduler() {
-			return new TaskSchedulerBuilder().build();
+		public TaskSchedulerWrapper taskScheduler() {
+			ThreadPoolTaskScheduler threadPoolTaskScheduler = new ThreadPoolTaskScheduler();
+
+			threadPoolTaskScheduler.setThreadNamePrefix(
+					"spring-cloud-kubernetes-ThreadPoolTaskScheduler-");
+			threadPoolTaskScheduler.setDaemon(true);
+
+			return new TaskSchedulerWrapper(threadPoolTaskScheduler);
 		}
 
 		/**

--- a/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/PollingConfigurationChangeDetector.java
+++ b/spring-cloud-kubernetes-config/src/main/java/org/springframework/cloud/kubernetes/config/reload/PollingConfigurationChangeDetector.java
@@ -16,6 +16,7 @@
 
 package org.springframework.cloud.kubernetes.config.reload;
 
+import java.time.Duration;
 import java.util.List;
 
 import javax.annotation.PostConstruct;
@@ -24,13 +25,15 @@ import io.fabric8.kubernetes.client.KubernetesClient;
 import org.apache.commons.logging.Log;
 import org.apache.commons.logging.LogFactory;
 
+import org.springframework.boot.task.TaskSchedulerBuilder;
 import org.springframework.cloud.kubernetes.config.ConfigMapPropertySource;
 import org.springframework.cloud.kubernetes.config.ConfigMapPropertySourceLocator;
 import org.springframework.cloud.kubernetes.config.SecretsPropertySource;
 import org.springframework.cloud.kubernetes.config.SecretsPropertySourceLocator;
 import org.springframework.core.env.AbstractEnvironment;
 import org.springframework.core.env.MapPropertySource;
-import org.springframework.scheduling.annotation.Scheduled;
+import org.springframework.scheduling.TaskScheduler;
+import org.springframework.scheduling.support.PeriodicTrigger;
 
 /**
  * A change detector that periodically retrieves secrets and configmaps and fire a reload
@@ -47,6 +50,11 @@ public class PollingConfigurationChangeDetector extends ConfigurationChangeDetec
 
 	private SecretsPropertySourceLocator secretsPropertySourceLocator;
 
+	private TaskScheduler taskExecutor;
+
+	private Duration period = Duration.ofMillis(1500);
+
+	@Deprecated
 	public PollingConfigurationChangeDetector(AbstractEnvironment environment,
 			ConfigReloadProperties properties, KubernetesClient kubernetesClient,
 			ConfigurationUpdateStrategy strategy,
@@ -56,15 +64,31 @@ public class PollingConfigurationChangeDetector extends ConfigurationChangeDetec
 
 		this.configMapPropertySourceLocator = configMapPropertySourceLocator;
 		this.secretsPropertySourceLocator = secretsPropertySourceLocator;
+		this.taskExecutor = new TaskSchedulerBuilder().build();
+	}
+
+	public PollingConfigurationChangeDetector(AbstractEnvironment environment,
+			ConfigReloadProperties properties, KubernetesClient kubernetesClient,
+			ConfigurationUpdateStrategy strategy,
+			ConfigMapPropertySourceLocator configMapPropertySourceLocator,
+			SecretsPropertySourceLocator secretsPropertySourceLocator,
+			TaskScheduler taskExecutor, ConfigReloadProperties configReloadProperties) {
+		super(environment, properties, kubernetesClient, strategy);
+
+		this.configMapPropertySourceLocator = configMapPropertySourceLocator;
+		this.secretsPropertySourceLocator = secretsPropertySourceLocator;
+		this.taskExecutor = taskExecutor;
+		this.period = configReloadProperties.getPeriod();
 	}
 
 	@PostConstruct
 	public void init() {
 		this.log.info("Kubernetes polling configuration change detector activated");
+		PeriodicTrigger trigger = new PeriodicTrigger(period.toMillis());
+		trigger.setInitialDelay(period.toMillis());
+		taskExecutor.schedule(this::executeCycle, trigger);
 	}
 
-	@Scheduled(initialDelayString = "${spring.cloud.kubernetes.reload.period:15000}",
-			fixedDelayString = "${spring.cloud.kubernetes.reload.period:15000}")
 	public void executeCycle() {
 
 		boolean changedConfigMap = false;

